### PR TITLE
fix(daemon): record agent CLI version in run diagnostics

### PR DESCRIPTION
**Pull request by libaojiang**
Refs #7609 (implements the first, per-run diagnostics half only)

## Why

Runtime detection already probes the selected agent executable for its exact CLI version, but ordinary runs did not freeze that value into the run record. As a result, terminal execution diagnostics reported `agent_cli_version_not_recorded`, making version-specific regressions difficult to correlate after the run had finished.

This keeps the change focused on per-run provenance. The separate stale-runtime Settings hint proposed in #7609 is intentionally out of scope.

## What users will see

Completed runs now report the selected runtime's detected CLI version under `executionDiagnostics.environment.agentCliVersion` instead of `not_collected` when a version is available. This applies to non-Codex runtimes such as Pi and OpenCode as well as existing runtime definitions.

## Surface area

- [ ] **UI** — new page / dialog / panel / menu item / setting / empty state in `apps/web` or `apps/desktop` (including Electron menu bar)
- [ ] **Keyboard shortcut** — new or changed
- [ ] **CLI / env var** — new `od` subcommand or flag, new `tools-dev` / `tools-pack` flag, or new `OD_*` env var
- [ ] **API / contract** — new `/api/*` endpoint, new SSE event, or changed shape in `packages/contracts`
- [ ] **Extension point** — new entry under `skills/`, `design-systems/`, `design-templates/`, or `craft/`, or change to the skills protocol
- [ ] **i18n keys** — added new translation keys (see `TRANSLATIONS.md` for the locale workflow)
- [ ] **New top-level dependency** — adding any new entry to the **root** `package.json` (`dependencies` or `devDependencies`); workspace-package `package.json` files are out of scope. Include a paragraph on what we get vs. what bytes we ship (see `CONTRIBUTING.md` → Code style)
- [ ] **Default behavior change** — changes what existing users experience without opting in (default model, default setting, file/SQLite schema, auto-network on startup, auto-install)
- [x] **None** — internal refactor, docs, tests, or translation update only

## Screenshots

Not applicable; this changes an existing diagnostics value and adds no UI.

## Bug fix verification

- Test path that reproduces the bug: `apps/daemon/tests/runtimes/run-runtime-version-provenance.test.ts`
- Did the test go red on `main` and green on this branch? **Yes.** On `main`, the real HTTP run returned `state: "not_collected"` with `missingReason: "agent_cli_version_not_recorded"`. On this branch it returns `state: "available"` with the exact fake OpenCode version `opencode-cli 9.8.7`.

## Validation

- `pnpm exec vitest run -c vitest.config.ts tests/runtimes/run-runtime-version-provenance.test.ts` — 1 passed
- `pnpm exec vitest run -c vitest.config.ts tests/runtimes/runs.test.ts` — 55 passed
- `pnpm --filter @open-design/daemon typecheck` — passed
- `pnpm --filter @open-design/daemon build` — passed
- `git diff --check` — passed
- `pnpm guard` — all checks reached before plugin-preview validation passed; the command could not complete in this partial clone because the intentionally omitted `plugins/` tree makes the preview manifest appear stale.

Local Windows note: the existing `runtime-version-provenance.test.ts` fixture resolves the locally installed Claude CLI instead of its extensionless fake executable on Windows (3 existing assertions fail with local `2.1.191` output). The new cross-platform OpenCode regression and the daemon checks above pass.

### CODE REVIEW

**Comment by lefarcen** [comment]
Thanks @libaojiang — scoping this to the per-run provenance half of #7609 is a nice way to keep the change reviewable. I’ve routed it for maintainer review and the usual PR triage path.

**Comment by lefarcen** [comment]
This touches run diagnostics that users read after a run, so it’ll get a manual QA pass before merge. Nothing needed from you right now — we’ll update here once that validation is done. Thanks for keeping the first slice focused.

**Review by nettee — COMMENTED** [review]
@libaojiang

Thanks for keeping this change focused on per-run provenance and for adding the real HTTP regression. I verified the live diff, the daemon shard covering the new test, and a clean diff check. I found two non-blocking preflight lifecycle and provenance edge cases in the new path; the details are inline.


<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

File: apps/daemon/src/server.ts [hunk 1]
```diff
@@ -12748,10 +12748,18 @@ export async function startServer({
...
-      // Optional argv flags are gated on the `--help` capability map, which used
-      // to be filled only by `GET /api/agents`. Probe it here so a daemon that
-      // has never served that route still builds the same argv as one that has.
-      await ensureDetectedRuntimeCapabilities(def.id, configuredAgentEnv);
+      // Freeze the selected CLI version on the Run and populate the optional
+      // argv capability map before spawn. Probe both in parallel so a daemon
+      // that has never served `GET /api/agents` has the same provenance and
+      // argv behavior as one whose detection cache is already warm.
+      const [runtimeVersions] = await Promise.all([
+        ensureDetectedRuntimeVersions(def.id, configuredAgentEnv),
```
Comment (nettee) [inline apps/daemon/src/server.ts]: Please make this probe use the already-captured agentLaunch (or make the newly resolved launch the one used for spawn). startChatRun resolves agentLaunch before the long setup above and later passes that object to spawn, but ensureDetectedRuntimeVersions calls resolveAgentLaunch again. A concurrent /api/agents detection can mark a broken PATH candidate unusable while this run is between those points, so the second resolution may probe candidate B while the run still starts candidate A. The run's stored version can then identify a CLI that never ran, which defeats the per-run provenance guarantee and makes diagnostics misleading. Please pass the captured launch/path into the version probe or update both buildArgs and spawn to use one authoritative resolution, and add a regression for that interleave.

<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

File: apps/daemon/src/server.ts [hunk 2]
```diff
@@ -12748,10 +12748,18 @@ export async function startServer({
...
+        ensureDetectedRuntimeCapabilities(def.id, configuredAgentEnv),
+      ]);
+      if (!run.preflightAgentCliVersion && runtimeVersions?.agentCliVersion) {
```
Comment (nettee) [inline apps/daemon/src/server.ts:12767]: Please add a cancellation check immediately after this new Promise.all and before mutating the run or calling buildArgs. cancel() finishes a queued run immediately when run.child is null; if Stop arrives while the --version probe is pending, this continuation resumes on a terminal run, persists preflightAgentCliVersion, and then calls buildArgs before the only later cancellation guard near line 13323. buildArgs is not universally pure (the Antigravity adapter writes settings.json), and the response and analytics emitted by cancel were built before this late mutation, so a later GET can disagree with the terminal result. Mirror the existing preflight cancellation cleanup branch here and add a delayed-probe cancellation regression.

<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

**Comment by lefarcen** [comment]
Thanks @libaojiang — @nettee's two inline notes are the main follow-ups on this head: guarding the new preflight path against cancellation races, and keeping the recorded CLI version tied to the same launch candidate the run actually spawns. Once those are addressed, the next pass here should be pretty small.

> 💡 To drive this PR to merge hands-free, paste this to your AI coding agent (Claude Code / Codex / opencode / Cursor …):
> `Take over nexu-io/open-design#7610 until it merges — read https://raw.githubusercontent.com/nexu-io/looper/main/skills/pr-takeover/SKILL.md and follow it.`

File: apps/daemon/src/server.ts (hunk 1 above)
Comment (libaojiang) [reply apps/daemon/src/server.ts]: Fixed in 5cc368e. startChatRun now passes its already-captured agentLaunch into both runtime version and capability preflight probes, so the probe/cache scope stays bound to the same selected and launch paths used for spawn. The regression captures candidate A, publishes the same unusable-path state a concurrent detection pass would, verifies fresh resolution moves to B, and then confirms the run still executes and records A.

File: apps/daemon/src/server.ts (hunk 2 above)
Comment (libaojiang) [reply apps/daemon/src/server.ts:12767]: Fixed in 5cc368e. A terminal/cancellation guard now runs immediately after the parallel probes and before version persistence or buildArgs; it mirrors the existing preflight cleanup and also releases the Antigravity model lock if held. The delayed-version regression cancels while the probe is blocked, then verifies the run keeps agentCliVersion uncollected and buildArgs is never called.

**Comment by libaojiang** [comment]
@nettee Both review follow-ups are addressed in 5cc368e and the two threads are resolved.

The new head passed all four daemon shards and E2E Vitest. The only failing leaf is UI P0 (entry-settings): the existing amr-onboarding.test.ts delayed-login case could not find its initial landing heading (86 other tests passed). The exact test passes locally with Chromium. I tried the failed-job rerun endpoint, but fork PR authors do not have the required Actions admin permission (403).

Could a maintainer rerun the failed UI job when convenient?

**Comment by lefarcen** [comment]
Reran the failed CI jobs on this head. Thanks for the detailed triage notes — that gives the next pass a clear baseline if the entry-settings leaf flakes again.

**Review by nettee — APPROVED** [review]
@libaojiang

I reviewed all three changed files and the launch, preflight, and cancellation paths. The captured launch is now shared by the version/capability probes and the eventual spawn, and the regression suite covers normal provenance, executable-fallback races, and cancellation. The focused runtime tests (3), adjacent runtime tests (59), daemon typecheck/build, and git diff check all pass. Nice work closing both lifecycle races while keeping the change focused.


<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

**Comment by lefarcen** [comment]
Queued for QA validation now that the current head has reviewer sign-off. We’ll use this pass to manually verify the runtime-version diagnostics path before merge.
